Section on cloning

JesperIRL opened the pull request:
Added a section on cloning from GitHub, and an example of building on Mac. Also fixed two minor bugs that caused problems with syntax highlighting in Atom.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Reviewers
 * [Stuart Marks](https://openjdk.java.net/census#smarks) (@stuart-marks - Committer) ⚠️ Review applies to 69d23dd647abc4c5d7741692821358df2bf6db06
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - no project role) ⚠️ Review applies to 3744213726c044f9b7859d218603e53a9f542e54
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - Author) ⚠️ Review applies to d43feb071c9191eca52be9b5412e1cdcc8d004ad
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - Author)
 * [Iris Clark](https://openjdk.java.net/census#iris) (@irisclark - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/guide pull/59/head:pull/59` \
`$ git checkout pull/59`

Update a local copy of the PR: \
`$ git checkout pull/59` \
`$ git pull https://git.openjdk.java.net/guide pull/59/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 59`

View PR using the GUI difftool: \
`$ git pr show -t 59`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/guide/pull/59.diff">https://git.openjdk.java.net/guide/pull/59.diff</a>

</details>

stuart-marks commented on src/index.md:
```diff
@@ -985,35 +985,109 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+    $ git clone git@github.com:JesperIRL/jdk.git
+    $ git remote add upstream git@github.com:openjdk/jdk.git
+
+In the example above I cloned my personal fork of the JDK mainline repository. You should of course use your own GitHub username instead. Then, by adding a new *remote* named 'upstream', we associate this clone with [openjdk/jdk](https://github.com/openjdk/jdk) as well. Doing this will allow the tooling to automatically create a PR request on [openjdk/jdk](https://github.com/openjdk/jdk) whenever ...
```
"PR request" => "PR"

JesperIRL replied:
Fixed.

stuart-marks commented on src/index.md:
```diff
@@ -985,35 +985,109 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+The recommendation is to always create a new branch for any change you intend to implement. By doing that you can easily work on many different changes in parallel in the same code repository. Unless you know what you are doing, the recommendation is also to always base your new branch on the `master` branch.
+
+    $ git checkout master
+    $ git checkout -b my-fix
```
I'm not sure whether we should recomment "git checkout -b" or "git switch -c". I think switch is safer, but the doc says that it's experimental....

JesperIRL replied:
Added a git switch example as well.

stuart-marks commented on src/index.md:
```diff
@@ -985,35 +985,109 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+  * It will look something like this: `ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIO8+egiIgWV+tE7LVVJmlR7WS2Lr3Fj7dXVo9HiasD6T openjdk-jdk`
+* Click "Add SSH key"
+
+Now you are ready to clone your [openjdk/jdk](https://github.com/openjdk/jdk) fork using SSH.
```
I suspect this SSH stuff is all fine for Linux and Mac. What about Windows? I imagine that one would need to install OpenSSH or something before this will work. I haven't done this though.

JesperIRL replied:
I've never done this on Windows. I'll try to get hold of a windows machine to try it out and add a windows section later.

stuart-marks commented on src/index.md:
```diff
@@ -985,35 +985,109 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+
+    $ curl https://download.java.net/java/GA/jdk16.0.1/7147401fd7354114ac51ef3e1328291f/9/GPL/openjdk-16.0.1_osx-x64_bin.tar.gz --output openjdk-16.0.1_osx-x64_bin.tar.gz
+    $ tar xzf openjdk-16.0.1_osx-x64_bin.tar.gz
+    $ # Get XCode from Appstore
```
I think this should be a separate text paragraph, not a comment in what looks like a cookbook script to building the JDK. Getting XCode is kind of a problem. You can get it from the App Store; but sometimes the App Store version is too new for the OS version you have, so you have to get an older XCode. But of course if XCode is too old then it can't build the JDK. Obviously we can't write a whole section that covers all these issues. Plus, it changes from time to time as Apple releases new versions of things. I'd suggest having a separate paragraph that talks about getting XCode from the App Store, or the possible need to get older versions from Apple's developer site. I would avoid putting too many details in this document. Probably the best fallback is to tell people to ask on the mailing list if they have trouble with this step.

JesperIRL replied:
I separated out the XCode info and put it in the text above, as a precondition to building.

kevinrushforth commented on src/index.md:
```diff
@@ -985,35 +985,106 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+
+The complete source code for the JDK is hosted at [GitHub](https://github.com). If you intend to make changes and contribute patches to the JDK, you should first fork the JDK repository on GitHub and clone your own fork as shown below. To fork a project on GitHub, go to the [project page](https://github.com/openjdk/jdk) and click the 'Fork' button in the upper right corner, then follow the on scr ...
+
+    $ git clone git@github.com:JesperIRL/jdk.git
+    $ git remote add upstream git@github.com:openjdk/jdk.git
```
I typically recommend using https for the "fetch" URL and "ssh" (i.e., `git@`) for the push URL. More importantly, I don't recommend using ssh at all for the upstream repo, since it is read-only for developers. Here is my suggestion:

```
$ git clone https://github.com/JesperIRL/jdk.git
$ cd jdk
$ git remote set-url origin --push git@github.com:JesperIRL/jdk.git
$ git remote add upstream https://github.com/openjdk/jdk.git
```

If you prefer not to introduce the complexity of separate fetch and push URLs for `origin`, I at least would like to see an `https` URL for the upstream.

JesperIRL replied:
How do we motivate recommending https for upstream? I mean it's not as if people can push regardless. Is there a git convention to use https for read only repos?

kevinrushforth replied:
GitHub defaults to https when you click on the green "Code" button to get the URL of a repo to clone. The Skara tooling also uses https URLs for openjdk/jdk in its examples (e.g., the example of fetching a pull request locally for testing). It seems better to suggest anonymous https for remotes that you don't need to push to (such as the upstream openjdk/jdk).

For the user's personal fork it really doesn't matter, so maybe it's better to drop that part of my suggestion (it's simpler too). So perhaps this?

```
$ git clone git@github.com:JesperIRL/jdk.git
$ cd jdk
$ git remote add upstream https://github.com/openjdk/jdk.git
```

JesperIRL replied:
Ok. Changed.

kevinrushforth commented on src/index.md:
```diff
@@ -985,35 +985,106 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+The recommendation is to always create a new branch for any change you intend to implement. By doing that you can easily work on many different changes in parallel in the same code repository. Unless you know what you are doing, the recommendation is also to always base your new branch on the `master` branch.
+
+    $ git checkout master
+    $ git checkout -b JDK-8272373
```
Minor: you could achieve the same thing with:

```
    $ git checkout -b JDK-8272373 master
```

JesperIRL replied:
Nice! Fixed.

kevinrushforth commented on src/index.md:
```diff
@@ -985,35 +985,111 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+
+## Generating an SSH key
+
+All pushes require an SSH key which must be installed on GitHub. If this is the first time you clone the [openjdk/jdk](https://github.com/openjdk/jdk) repository you may want to create an SSH key to use with it. For security reasons you should always create new keys and use different keys with each repository you clone. The `ssh-keygen` command generates an SSH key. The `-t` option determines whi ...
```
Maybe "if this is the first time you clone _your personal fork of_ the openjdk/jdk repo ..."

JesperIRL replied:
Fixed.

JesperIRL commented:
Thank you @kevinrushforth for your review

dholmes-ora commented:
Hi Jesper,

There seems to be quite a bit of overlap/duplication with what can be found on the skara wiki. I don't think it makes sense to duplicate technical details, especially when so much is subjective (http vs. https vs. ssh). I'd rather see links to that wiki and the wiki itself updated if needed.

Cheers,
David

JesperIRL commented:
I tried with a newly created GitHub account and verified that the SSH key is required already when cloning. So I added an initial clone example that uses https. Also added references to the Skara documentation.

dholmes-ora commented on src/index.md:
```diff
@@ -985,35 +985,117 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+
+    $ git clone https://github.com/openjdk/jdk.git
+
+If you intend to contribute patches to the JDK, you should first *fork* the JDK repository on GitHub and clone your own fork as shown below. To fork a project on GitHub, go to the [project page](https://github.com/openjdk/jdk) and click the 'Fork' button in the upper right corner, then follow the on screen instructions.
```
s/own fork/own personal fork/

This introduces the "personal fork" terminology.

JesperIRL replied:
Fixed.

dholmes-ora commented on src/index.md:
```diff
@@ -985,35 +985,117 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+
+If you intend to contribute patches to the JDK, you should first *fork* the JDK repository on GitHub and clone your own fork as shown below. To fork a project on GitHub, go to the [project page](https://github.com/openjdk/jdk) and click the 'Fork' button in the upper right corner, then follow the on screen instructions.
+
+All pushes to [openjdk/jdk](https://github.com/openjdk/jdk) require an SSH key which must be installed on GitHub. If this is the first time you clone your personal fork of the [openjdk/jdk](https://github.com/openjdk/jdk) repository you may want to create an SSH key to use with it. See [Generating an SSH key] below. Once you have your private fork and an SSH key to go with it, go ahead and clone.
```
s/private fork/personal fork/

JesperIRL replied:
Fixed.

dholmes-ora commented on src/index.md:
```diff
@@ -985,35 +985,117 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+    $ cd jdk
+    $ git remote add upstream https://github.com/openjdk/jdk.git
+
+In the example above Duke cloned his personal fork of the JDK mainline repository using SSH. You should of course use your own GitHub username instead. Then, by adding a new *remote* named 'upstream', the clone is associated with [openjdk/jdk](https://github.com/openjdk/jdk). Doing this will allow the tooling to automatically create a PR on [openjdk/jdk](https://github.com/openjdk/jdk) whenever a ...
```
s/private fork/personal fork/

JesperIRL replied:
Fixed.

dholmes-ora reviewed (approved):
Hi Jesper,

A couple of minor nits but otherwise seems a reasonable balance of information to get someone started.

Thanks,
David

JesperIRL commented:
Thank you for your review David!

irisclark commented on src/index.md:
```diff
@@ -1437,79 +1519,6 @@ Option 2: _**`hg`**_ `fetch`
-    |      o . .o..   |
-    +-----------------+
-
-The `~/.ssh/id_rsa.pub` is a text file containing the public ssh key. This file should be mailed as an attachment along with the JDK username to [keys(at)openjdk.java.net](mailto:keys-at-openjdk.java.net). An administrator will install your key on the server and notify you on completion. This process may take a couple of days.
```
I'm not sure how to handle this.  All of the new section is about using Git; however, these instructions for sending to keys@ojn are still needed for the very few Projects which still develop using hg.ojn or for those intending to use cr.ojn.  I referenced them in the past few weeks when helping a user so I'm reluctant to completely eliminate these few sentences quite yet.

magicus replied:
Can we keep a section "How to do development on the legacy Mercurial server"?

JesperIRL replied:
I created a "Working with the legacy Mercurial servers" section and pulled all the mercurial stuff from "Repositories" and "Producing a Changeset" there. Some sentences was placed in other parts of the guide where I felt they were more in context. This made the change significantly larger but I also think the result is a lot better and made the guide as a whole more up to date.

magicus commented on src/index.md:
```diff
@@ -985,35 +985,117 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+
+If you intend to contribute patches to the JDK, you should first *fork* the JDK repository on GitHub and clone your own *personal fork* as shown below. To fork a project on GitHub, go to the [project page](https://github.com/openjdk/jdk) and click the 'Fork' button in the upper right corner, then follow the on screen instructions.
+
+All pushes to [openjdk/jdk](https://github.com/openjdk/jdk) require an SSH key which must be installed on GitHub. If this is the first time you clone your personal fork of the [openjdk/jdk](https://github.com/openjdk/jdk) repository you may want to create an SSH key to use with it. See [Generating an SSH key] below. Once you have your personal fork and an SSH key to go with it, go ahead and clone ...
```
This comment is either wrong or irrelevant. A normal developer cannot push to the openjdk/jdk repo. For your personal fork, you may chose to not use SSH, and instead use https.

JesperIRL replied:
Rephrased and clarified. I hope :-)

magicus commented on src/index.md:
```diff
@@ -985,35 +985,117 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+
+Here we create a new branch called `JDK-8272373` based on the `master` branch and set the repository up to work in that new branch.
+
+Starting from Git version 2.23 there's also `git switch` that can be used instead of `git checkout`.
```
Since 'git checkout' has the potential for catastrophic data loss, all modern git usage suggest using 'git switch' for switching branches. I recommend you suggest just 'git switch' and then perhaps parenthetically notices that prior to git 2.23 'git checkout' is needed instead.

JesperIRL replied:
Removed the checkout example.

magicus commented on src/index.md:
```diff
@@ -985,35 +985,117 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+
+## Generating an SSH key
+
+For security reasons you should always create new keys and use different keys with each repository you clone. The `ssh-keygen` command generates an SSH key. The `-t` option determines which type of key to create. `ed25519` is recommended. `-C` is used to add a comment in the key file, to help you remember which key it is. While it’s possible to use SSH without a passphrase, this is **strongly dis ...
```
Personally, I believe using ssh is a lot of hassle (as these paragraphs shows; a third of the new section is devoted to ssh key handling) and think the https/PAT route is much simpler. Maybe you should at least mention it as an alternative, if you do not want to suggest it as the primary way to access Github.

JesperIRL replied:
I have added a comment about https. I prefer to keep ssh for the most part since that's how I work and I can't verify my examples if they should assume that there are no ssh key installed.

magicus commented on src/index.md:
```diff
@@ -985,35 +985,117 @@ For the purposes of brevity this document will use the term "bug" to refer to bo
+The built JDK can be found in `build/linux-x86_64-server-release/jdk`. The exact path depends on your build platform and selected configuration.
+
+The second example is from a clean (newly installed) Mac running MacOS Big Sur. Please note that in this case there are some steps taken outside of the terminal. First XCode and the XCode command line tools must be installed. It could be that the most recent version of XCode that you get from App Store is too new to have been properly tested with the JDK build. See [the JDK build instructions](ht ...
+In this example [Mac Ports](https://www.macports.org) is used to install `autoconf`. `autoconf` can also be installed using [Homebrew](https://brew.sh) and surely through other sources as well.
```
Mac Ports is virtually dead. Homebrew usage dwarfs Mac Ports by magnitudes (see https://www.scivision.dev/homebrew-macports-fink/ for usage data). I recommend you switch the example to use Homebrew (and possibly mentions Mac Ports, but seriously, I don't think anyone ever tried to build OpenJDK on mac using Mac Ports so I'm not sure how well that woks).

JesperIRL replied:
I use Mac Ports. It works fine (as the example shows) to build the JDK with it. Since I don't have Homebrew installed I can't write examples using it.

magicus commented:
@JesperIRL Did you give up on this? I think it's basically sound but need some additional fixes.

JesperIRL commented:
:-) No, I haven't given up. I was on vacation when these comments came in, and then I have been occupied with other work. I will dig into this asap.

JesperIRL commented:
@irisclark @magicus  Thank you for your reviews!  The new version should address all your concerns in one way or another :-)

magicus reviewed (approved):
It's a lot to take in, and I've just skimmed parts of it, but I think it looks good. In any way, it's way better than what was before and if something turns out to be not perfect, we can fix it later.

JesperIRL commented:
/integrate
